File: 229-majority-element-ii/1-6-2019.java
class Solution {
    public List<Integer> majorityElement(int[] nums) {
        // The list will only have at most 2 element
        if (nums == null || nums.length == 0) {
            return new ArrayList<>();
        }
        int count1 = 0;
        int count2 = 0;
        int n1 = 0;
        int n2 = 0;
        for (int num : nums) {
            if (n1 == num) {
                count1++;
            } else if (n2 == num) {
                count2++;
            } else if (count1 == 0) {
                n1 = num;
                count1++;
            } else if (count2 == 0) {
                n2 = num;
                count2++;
            } else {
                count1--;
                count2--;
            }
        }
        List<Integer> res = new ArrayList<>();
        int n = nums.length;
        count1 = count2 = 0;
        for (int num : nums) {
            if (num == n1) {
                count1++;
            } else if (num == n2) {
                count2++;
            }
        }
        if (count1 > n / 3) {
            res.add(n1);
        } 
        if (count2 > n / 3) {
            res.add(n2);
        }
        return res;
    }
}

File: 229-majority-element-ii/1-6-2019.py
class Solution(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        candidate1, count1 = None, 0
        candidate2, count2 = None, 0
        for num in nums:
            if num == candidate1:
                count1 += 1
            elif num == candidate2:
                count2 += 1
            elif count1 == 0:
                candidate1 = num
                count1 += 1
            elif count2 == 0:
                candidate2 = num
                count2 += 1
            else:
                count1 -= 1
                count2 -= 1
        count1, count2 = 0, 0
        for num in nums:
            if candidate1 == num:
                count1 += 1
            if candidate2 == num:
                count2 += 1
        n = len(nums)
        if count1 > n / 3 and count2 > n / 3:
            return [candidate1, candidate2]
        elif count1 > n / 3:
            return [candidate1]
        elif count2 > n / 3:
            return [candidate2]
        else:
            return []
